feat: Notifications Inbox Screen

HamzaIsrar12 opened the pull request:
### Description:

[LEARNER-10346](https://2u-internal.atlassian.net/browse/LEARNER-10346)

A Notification Inbox Screen that organizes notifications into three categories: **Recent** (last 24 hours), **This Week** (past 7 days), and **Older**.

### Screenshots:

| Inbox Screen | Empty Screen | Error Screen|
|--------|--------|--------|
| ![Screenshot_20250108_144055](https://github.com/user-attachments/assets/80515438-666a-492f-b444-18d6709fb190) | ![No Notifications Dark](https://github.com/user-attachments/assets/c1542ed6-aff1-48ce-bcd1-bbcab0c60eff) | ![Error Notifications Dark](https://github.com/user-attachments/assets/9a47b256-7f24-4bb7-a1da-b1c4bc3df4ca) |
| ![Screenshot_20250108_144151](https://github.com/user-attachments/assets/10a8f5b8-4504-49f1-b1e3-a081b9763c30) | ![No Notifications Light](https://github.com/user-attachments/assets/a4b71e56-c7dd-4aee-afe8-4c8d8accf90f) | ![Error Notifications Light](https://github.com/user-attachments/assets/7590bb2a-3e2b-4e5e-a540-eac8dc5c4001) |

### Figma:

[Mobile Notifications UI Screen](https://www.figma.com/design/I9xgrJzdTje5vKgh6DD57C/Mobile-UI-Screens-(2025)?node-id=1-4933&t=Q3n6mSTW27u9mk0t-0)

HamzaIsrar12 committed:
chore: Base Inbox View

Fixes: LEARNER-10346

HamzaIsrar12 committed:
chore: Prepare the Data and Domain Layer

Fixes: LEARNER-10346

HamzaIsrar12 committed:
chore: Group notifications into sections

Fixes: LEARNER-10346

HamzaIsrar12 committed:
feat: Show notifications in the Inbox

Fixes: LEARNER-10346

HamzaIsrar12 committed:
feat: Add Inbox States

Fixes: LEARNER-10346

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/data/api/APIConstants.kt:
```diff
@@ -2,4 +2,5 @@ package org.openedx.notifications.data.api
 
 object APIConstants {
     const val NOTIFICATION_COUNT = "/api/notifications/count/"
+    const val NOTIFICATIONS_INBOX = "api/notifications/"
```
For APIs starting with `/`, it is recommended to follow the common practice for defining API paths.

k1rill replied:
NOTIFICATION_COUNT begins from slash, but NOTIFICATIONS_INBOX not, is all correct here?

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/data/api/NotificationsApi.kt:
```diff
@@ -1,9 +1,17 @@
 package org.openedx.notifications.data.api
 
 import org.openedx.notifications.data.model.NotificationsCountResponse
+import org.openedx.notifications.data.model.InboxNotificationsResponse
```
required auto-format.

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/data/model/InboxNotificationsResponse.kt:
```diff
@@ -0,0 +1,110 @@
+    )
+}
+
+data class NotificationContent(
+    @SerializedName("p") val paragraph: String,
+    @SerializedName("strong") val strongText: String,
+    @SerializedName("topic_id") val topicId: String?,
+    @SerializedName("parent_id") val parentId: String?,
+    @SerializedName("thread_id") val threadId: String?,
+    @SerializedName("comment_id") val commentId: String?,
+    @SerializedName("post_title") val postTitle: String,
+    @SerializedName("course_name") val courseName: String,
+    @SerializedName("replier_name") val replierName: String,
+    @SerializedName("email_content") val emailContent: String,
```
[Optional] The following attributes are missing from the data model.
- `author_name`
- `author_pronoun`

HamzaIsrar12 replied:
I'm not seeing these attributes in my api response. Please dm the username/pass so I can verify it.

farhan-arshad-dev replied:
shared on Slack.

HamzaIsrar12 replied:
Fields added.

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/data/model/InboxNotificationsResponse.kt:
```diff
@@ -0,0 +1,110 @@
+            },
+            InboxSection.THIS_WEEK to notifications.filter {
+                val createdTime = it.created?.time ?: 0L
+                createdTime in (weekThresholdMillis + 1) until recentThresholdMillis
```
use case to add +1 with `weekThresholdMillis`?

HamzaIsrar12 replied:
You are right, we don't need it here because of the less than operator in Older section.

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/data/repository/NotificationsRepository.kt:
```diff
@@ -1,10 +1,18 @@
+
+    suspend fun getInboxNotifications(page: Int): InboxNotifications {
+        return api.getInboxNotifications(
+            appName = "discussion",
```
IMO, better the define the `discussion` in APIContant.

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/presentation/inbox/InboxUIState.kt:
```diff
@@ -0,0 +1,15 @@
+
+sealed class InboxUIState {
+
+    data class Data(
+        val notifications: Map<InboxSection, List<NotificationItem>>,
+    ) : InboxUIState()
```
can be a single linner.

HamzaIsrar12 replied:
We may need to add more in the future, so I prefer to leave it as it is for now.

k1rill commented on core/src/edx/org/openedx/core/ui/theme/Colors.kt:
```diff
@@ -90,6 +90,7 @@ val light_progress_bar_background_color = Color(0xFFE7E4DB)
 val light_primary_card_caution_background = Color(0xFFF3F1ED)
 val light_primary_card_info_background = Color(0xFFE7E4DB)
 
+val light_notification_timestamp = Color(0xFF707070)
```
Strange to see "timestamp" in the color naming. Timestamp isn't in use on the presentation level, maybe you can rename it?

k1rill commented on notifications/src/main/java/org/openedx/notifications/presentation/inbox/NotificationsInboxFragment.kt:
```diff
@@ -0,0 +1,483 @@
+}
+
+@Composable
+fun InboxStateView(
```
maybe it should be private also?

k1rill reviewed (commented):
Good in general, just some questions

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/utils/TextUtils.kt:
```diff
@@ -0,0 +1,111 @@
+
+object TextUtils {
+
+    fun dateToRelativeTimeString(context: Context, date: Date?): String {
```
there is a method `DateUtils.getRelativeTimeSpanString`, we can explore to achieve the requirement.

HamzaIsrar12 replied:
I never knew such function existed but our requirements is a bit different from the outcome of the function i.e we don't want the "ago" text.

farhan-arshad-dev replied:
for now, we can move forward with the current implementation.

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/presentation/inbox/NotificationsInboxFragment.kt:
```diff
@@ -0,0 +1,483 @@
+        setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+        setContent {
+            OpenEdXTheme {
+                val windowSize = rememberWindowSize()
```
Are we accounting for the tablet view as part of this PR?

HamzaIsrar12 replied:
Yes

farhan-arshad-dev replied:
Then need to improve the design for the tablet view.
![image](https://github.com/user-attachments/assets/29dc7fcb-a648-4580-bb00-4003adfceeb6)

HamzaIsrar12 replied:
Fixed!

![Screenshot_20250107_144349](https://github.com/user-attachments/assets/88e56ef3-86f1-49f5-8bf7-5ce35f32dd2d)

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/presentation/inbox/NotificationsInboxFragment.kt:
```diff
@@ -0,0 +1,483 @@
+    ) {
+        Text(
+            text = stringResource(id = section.titleResId),
+            modifier = Modifier.padding(bottom = 24.dp),
```
IMO, better to place the padding in the parent component.

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/presentation/inbox/NotificationsInboxFragment.kt:
```diff
@@ -0,0 +1,483 @@
+    uiState: InboxUIState,
+    onReloadNotifications: () -> Unit = { },
+) {
+    val iconResId = if (uiState is InboxUIState.Empty) Icons.Outlined.Notifications
+    else Icons.Outlined.SignalWifiStatusbarConnectedNoInternet4
```
The icon color didn't match with the [Figma designs](https://www.figma.com/design/I9xgrJzdTje5vKgh6DD57C/Mobile-UI-Screens-(2025)?node-id=116-3032&t=EnpEWGB5oFiTm5hh-0)

HamzaIsrar12 replied:
I know that. I intentionally used the color already used in the app so we wouldn't have to introduce new colors.

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/presentation/inbox/NotificationsInboxFragment.kt:
```diff
@@ -0,0 +1,483 @@
+        )
+        Spacer(modifier = Modifier.height(12.dp))
+
+        Text(
+            text = stringResource(descriptionResId),
+            style = MaterialTheme.appTypography.bodyLarge,
+            color = MaterialTheme.appColors.textPrimary,
+            textAlign = TextAlign.Center,
+        )
```
As per the [Figma designs](https://www.figma.com/design/I9xgrJzdTje5vKgh6DD57C/Mobile-UI-Screens-(2025)?node-id=116-3032&t=EnpEWGB5oFiTm5hh-0) the font size should be 18.sp

HamzaIsrar12 replied:
Yes. But our app doesn't have that size, and it would be outside our design guidelines. I used 16.sp for consistency.

farhan-arshad-dev commented on notifications/src/main/java/org/openedx/notifications/presentation/inbox/NotificationsInboxFragment.kt:
```diff
@@ -0,0 +1,483 @@
+        )
+        Spacer(modifier = Modifier.height(12.dp))
+
+        if (uiState is InboxUIState.Error) {
+            OpenEdXPrimaryButton(
+                modifier = Modifier
+                    .widthIn(Dp.Unspecified, 162.dp),
+                text = stringResource(id = coreR.string.core_reload),
+                textColor = MaterialTheme.appColors.secondaryButtonText,
+                backgroundColor = MaterialTheme.appColors.secondaryButtonBackground,
+                onClick = onReloadNotifications,
+            )
+        }
```
According to the [Figma design](https://www.figma.com/design/I9xgrJzdTje5vKgh6DD57C/Mobile-UI-Screens-(2025)?node-id=116-3106&t=EnpEWGB5oFiTm5hh-0), no button has been placed.

HamzaIsrar12 replied:
The button is present in the light mode, Gabriel probably forgot about it in the dark mode.

farhan-arshad-dev reviewed (commented):
Some minor concerns.

- Text color doesn't match with the [Figma designs](https://www.figma.com/design/I9xgrJzdTje5vKgh6DD57C/Mobile-UI-Screens-(2025)?node-id=1-4933&t=EnpEWGB5oFiTm5hh-0) of static string like `responded to a post...`.
- Pull to refresh should be implemented.
- No Quotation marks around the Post title.
- bold Text weight looks higher than the [Figma designs](https://www.figma.com/design/I9xgrJzdTje5vKgh6DD57C/Mobile-UI-Screens-(2025)?node-id=1-4933&t=EnpEWGB5oFiTm5hh-0).
- `Ago` word needs to be removed after the date as per the [Fimga designs](https://www.figma.com/design/I9xgrJzdTje5vKgh6DD57C/Mobile-UI-Screens-(2025)?node-id=1-4933&t=EnpEWGB5oFiTm5hh-0).
- Sapce should be higher between notification text and date.

HamzaIsrar12 committed:
refactor: Minor improvements

Fixes: LEARNER-10346

HamzaIsrar12 commented:
> Text color doesn't match with the Figma designs of static string like "responded to a post".  
- As discussed with Moiz, we should adopt the colors from our design palette.

> Pull to refresh should be implemented.  
- This will be done in a separate PR. However, I already have the code ready and can push it if needed.

> No quotation marks around the Post title.  
- Added.

> Bold text weight appears heavier than in the Figma designs.  
- Changed from 700 to 600. Note that Figma used 500, but since we are using the same color, 500 didn't look as good.

> "Ago" should be removed after the date as per the Figma designs.  
- Removed.

> Space between the notification text and date should be higher.  
- The space is already set to 8.dp.

HamzaIsrar12 committed:
refactor: Update data model and fix UI for tablet

Fixes: LEARNER-10345

farhan-arshad-dev reviewed (commented):
Comments added for the pending points.

farhan-arshad-dev reviewed (approved):
LGTM 🔥 .

k1rill reviewed (approved):
LGTM
